crates/wasm-encoder/src/reencode.rs

```diff
@@ -528,6 +684,261 @@ impl Reencode for RoundtripReencoder {
+                    break;
+                }
+                wasmparser::Payload::ModuleSection {
+                    parser: subparser,
+                    unchecked_range,
+                } => {
+                    handle_intersperse_section_hook(
+                        reencoder,
+                        component,
+                        &mut last_section,
+                        Some(crate::ComponentSectionId::CoreModule),
+                    )?;
+                    let mut module = crate::Module::new();
+                    offset += unchecked_range.len();
+                    parse_core_module(reencoder, &mut module, subparser, &data[unchecked_range])?;
+                    component.section(&crate::ModuleSection(&module));
+                }
+                wasmparser::Payload::ComponentSection {
+                    parser: subparser,
+                    unchecked_range,
+                } => {
+                    handle_intersperse_section_hook(
+                        reencoder,
+                        component,
+                        &mut last_section,
+                        Some(crate::ComponentSectionId::Component),
+                    )?;
+                    let mut nested_component = crate::Component::new();
+                    offset += unchecked_range.len();
+                    parse_component(
+                        reencoder,
+                        &mut nested_component,
+                        subparser,
+                        &data[unchecked_range],
+                    )?;
+                    component.section(&crate::NestedComponentSection(&nested_component));
+                }
+                wasmparser::Payload::InstanceSection(section) => {
+                    handle_intersperse_section_hook(
+                        reencoder,
+                        component,
+                        &mut last_section,
+                        Some(crate::ComponentSectionId::CoreInstance),
+                    )?;
+                    let mut is = crate::InstanceSection::new();
+                    for inst in section {
+                        match inst? {
+                            wasmparser::Instance::Instantiate { module_index, args } => {
+                                let aa = args
+                                    .iter()
+                                    .map(|a| (a.name, crate::ModuleArg::Instance(a.index)));
+                                is.instantiate(module_index, aa);
+                            }
+                            wasmparser::Instance::FromExports(exports) => {
+                                let ee = exports
+                                    .iter()
+                                    .map(|e| (e.name, reencoder.export_kind(e.kind), e.index));
+                                is.export_items(ee);
+                            }
+                        };
+                    }
+                    component.section(&is);
```

Comment:
Mind splitting these section handlers into components such as:

* A `parse_instance_section` method which takes the wasmparser-based reader
* A `From` and conversion method such as `parse_instance` for each item in the section

That'll help keep these sorts of impls a bit neater and additionally enable implementors of `Reencode` to be able to customize the smallest portion available (and or call the "biggest" portion they want)